### ISSUE TITLE
Fails to compile correctly with -flto

### ISSUE THREAD

**Issue by kentfredric** [ajgb/crypt-curve25519#6]
```
# perl Makefile.PL OPTIMIZE="-O3 -flto"
Checking if your kit is complete...
Looks good
MakeMaker (v7.24)
Warning (non-fatal): Target 'dynamic' depends on targets in skipped section 'dynamic_lib'
Warning (non-fatal): Target 'static' depends on targets in skipped section 'static_lib'
Generating a Unix-style Makefile
Writing Makefile for Crypt::Curve25519::curve25519donna
Generating a Unix-style Makefile
Writing Makefile for Crypt::Curve25519
Writing MYMETA.yml and MYMETA.json
# make
cp lib/Crypt/Curve25519.pm blib/lib/Crypt/Curve25519.pm
AutoSplitting blib/lib/Crypt/Curve25519.pm (blib/lib/auto/Crypt/Curve25519)
cd curve25519-donna-master && make LIBPERL_A="libperl.a" LINKTYPE="dynamic" OPTIMIZE="-O3 -flto" PREFIX="/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot" PASTHRU_DEFINE=' ' PASTHRU_INC='-I. '
make[1]: Entering directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
cc -c -I.   -fstack-protector-strong -fno-stack-protector -DPERL_HASH_FUNC_SDBM -DPERL_DISABLE_PMC -fwrapv -fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O3 -flto   -DVERSION=\"\" -DXS_VERSION=\"\" -fPIC "-I/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/CORE"   curve25519-donna-c64.c
ar cr curve25519-donna-c64.a curve25519-donna-c64.o
ar: curve25519-donna-c64.o: plugin needed to handle lto object
: curve25519-donna-c64.a
make[1]: Leaving directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
make[1]: Entering directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
Running Mkbootstrap for Curve25519 ()
chmod 644 "Curve25519.bs"
"/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/bin/perl5.25.10" -MExtUtils::Command::MM -e 'cp_nonempty' -- Curve25519.bs blib/arch/auto/Crypt/Curve25519/Curve25519.bs 644
"/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/bin/perl5.25.10" "/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/ExtUtils/xsubpp"  -typemap '/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/ExtUtils/typemap'  Curve25519.xs > Curve25519.xsc
mv Curve25519.xsc Curve25519.c
cc -c  -I. -fstack-protector-strong -fno-stack-protector -DPERL_HASH_FUNC_SDBM -DPERL_DISABLE_PMC -fwrapv -fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O3 -flto   -DVERSION=\"0.05\" -DXS_VERSION=\"0.05\" -fPIC "-I/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/CORE"   Curve25519.c
rm -f blib/arch/auto/Crypt/Curve25519/Curve25519.so
cc  -shared  -fstack-protector-strong -fno-stack-protector -O3 -march=native -mtune=native -L/usr/local/lib Curve25519.o  -o blib/arch/auto/Crypt/Curve25519/Curve25519.so curve25519-donna-master/curve25519-donna-c64.a \
      \
  
chmod 755 blib/arch/auto/Crypt/Curve25519/Curve25519.so
# make test
"/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/bin/perl5.25.10" -MExtUtils::Command::MM -e 'cp_nonempty' -- Curve25519.bs blib/arch/auto/Crypt/Curve25519/Curve25519.bs 644
make[1]: Entering directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
make[1]: Entering directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
No tests defined for Crypt::Curve25519::curve25519donna extension.
make[1]: Leaving directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
PERL_DL_NONLAZY=1 "/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/bin/perl5.25.10" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/00-compile.t ........ 1/1 
#   Failed test 'Crypt/Curve25519.pm loaded ok'
#   at t/00-compile.t line 37.
#          got: '512'
#     expected: '0'
Can't load '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so' for module Crypt::Curve25519: /home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so: undefined symbol: curve25519_donna at /home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/DynaLoader.pm line 193.
 at -e line 1.
Compilation failed in require at -e line 1.
# Looks like you failed 1 test of 1.
t/00-compile.t ........ Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
t/01-exceptions.t ..... 1/? 
#   Failed test 'use Crypt::Curve25519;'
#   at t/01-exceptions.t line 3.
#     Tried to use 'Crypt::Curve25519'.
#     Error:  Can't load '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so' for module Crypt::Curve25519: /home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so: undefined symbol: curve25519_donna at /home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/DynaLoader.pm line 193.
#  at t/01-exceptions.t line 3.
# Compilation failed in require at t/01-exceptions.t line 3.
# BEGIN failed--compilation aborted at t/01-exceptions.t line 3.

#   Failed test 'curve25519(): Using primitive function requires two arguments'
#   at t/01-exceptions.t line 8.
#                   'Undefined subroutine &Crypt::Curve25519::curve25519 called at t/01-exceptions.t line 6.
# '
#     doesn't match '(?^:Using primitive function requires two arguments)'

#   Failed test 'curve25519_secret_key(): Secret key requires 32 bytes'
#   at t/01-exceptions.t line 13.
#                   'Undefined subroutine &main::curve25519_secret_key called at t/01-exceptions.t line 11.
# '
#     doesn't match '(?^:Secret key requires 32 bytes)'

#   Failed test 'curve25519_public_key(): Secret key requires 32 bytes'
#   at t/01-exceptions.t line 18.
#                   'Undefined subroutine &main::curve25519_public_key called at t/01-exceptions.t line 16.
# '
#     doesn't match '(?^:Secret key requires 32 bytes)'

#   Failed test 'curve25519_shared_secret:() Calculating shared secret requires public key'
#   at t/01-exceptions.t line 23.
#                   'Undefined subroutine &main::curve25519_shared_secret called at t/01-exceptions.t line 21.
# '
#     doesn't match '(?^:Calculating shared secret requires public key)'

#   Failed test 'curve25519_public_key(): Basepoint requires 32 bytes'
#   at t/01-exceptions.t line 29.
#                   'Undefined subroutine &main::curve25519_public_key called at t/01-exceptions.t line 27.
# '
#     doesn't match '(?^:Basepoint requires 32 bytes)'

#   Failed test 'curve25519_shared_secret(): Public key requires 32 bytes'
#   at t/01-exceptions.t line 34.
#                   'Undefined subroutine &main::curve25519_shared_secret called at t/01-exceptions.t line 32.
# '
#     doesn't match '(?^:Public key requires 32 bytes)'
# Looks like you failed 7 tests of 7.
t/01-exceptions.t ..... Dubious, test returned 7 (wstat 1792, 0x700)
Failed 7/7 subtests 
t/02-synopsis-proc.t .. 1/? 
#   Failed test 'use Crypt::Curve25519;'
#   at t/02-synopsis-proc.t line 3.
#     Tried to use 'Crypt::Curve25519'.
#     Error:  Can't load '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so' for module Crypt::Curve25519: /home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so: undefined symbol: curve25519_donna at /home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/DynaLoader.pm line 193.
#  at t/02-synopsis-proc.t line 3.
# Compilation failed in require at t/02-synopsis-proc.t line 3.
# BEGIN failed--compilation aborted at t/02-synopsis-proc.t line 3.
Undefined subroutine &main::curve25519_secret_key called at t/02-synopsis-proc.t line 6.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 1.
t/02-synopsis-proc.t .. Dubious, test returned 255 (wstat 65280, 0xff00)
Failed 1/1 subtests 
t/03-synopsis-oo.t .... 1/? 
#   Failed test 'use Crypt::Curve25519;'
#   at t/03-synopsis-oo.t line 3.
#     Tried to use 'Crypt::Curve25519'.
#     Error:  Can't load '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so' for module Crypt::Curve25519: /home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so: undefined symbol: curve25519_donna at /home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/DynaLoader.pm line 193.
#  at t/03-synopsis-oo.t line 3.
# Compilation failed in require at t/03-synopsis-oo.t line 3.
# BEGIN failed--compilation aborted at t/03-synopsis-oo.t line 3.
Undefined subroutine &Crypt::Curve25519::curve25519_secret_key called at /home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/lib/Crypt/Curve25519.pm line 41.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 1.
t/03-synopsis-oo.t .... Dubious, test returned 255 (wstat 65280, 0xff00)
Failed 1/1 subtests 
t/04-proc-vs-oo.t ..... 1/? 
#   Failed test 'use Crypt::Curve25519;'
#   at t/04-proc-vs-oo.t line 4.
#     Tried to use 'Crypt::Curve25519'.
#     Error:  Can't load '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so' for module Crypt::Curve25519: /home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so: undefined symbol: curve25519_donna at /home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/DynaLoader.pm line 193.
#  at t/04-proc-vs-oo.t line 4.
# Compilation failed in require at t/04-proc-vs-oo.t line 4.
# BEGIN failed--compilation aborted at t/04-proc-vs-oo.t line 4.
Undefined subroutine &Crypt::Curve25519::curve25519_secret_key called at /home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/lib/Crypt/Curve25519.pm line 41.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 1.
t/04-proc-vs-oo.t ..... Dubious, test returned 255 (wstat 65280, 0xff00)
Failed 1/1 subtests 
t/05-primitive.t ...... 1/? 
#   Failed test 'use Crypt::Curve25519;'
#   at t/05-primitive.t line 3.
#     Tried to use 'Crypt::Curve25519'.
#     Error:  Can't load '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so' for module Crypt::Curve25519: /home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so: undefined symbol: curve25519_donna at /home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/DynaLoader.pm line 193.
#  at t/05-primitive.t line 3.
# Compilation failed in require at t/05-primitive.t line 3.
# BEGIN failed--compilation aborted at t/05-primitive.t line 3.
Undefined subroutine &main::curve25519_secret_key called at t/05-primitive.t line 11.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 1.
t/05-primitive.t ...... Dubious, test returned 255 (wstat 65280, 0xff00)
Failed 1/1 subtests 
t/Crypt-Curve25519.t .. Can't load '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so' for module Crypt::Curve25519: /home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/blib/arch/auto/Crypt/Curve25519/Curve25519.so: undefined symbol: curve25519_donna at /home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/DynaLoader.pm line 193.
 at t/Crypt-Curve25519.t line 3.
Compilation failed in require at t/Crypt-Curve25519.t line 3.
BEGIN failed--compilation aborted at t/Crypt-Curve25519.t line 3.
t/Crypt-Curve25519.t .. Dubious, test returned 2 (wstat 512, 0x200)
No subtests run 

Test Summary Report
-------------------
t/00-compile.t      (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
t/01-exceptions.t   (Wstat: 1792 Tests: 7 Failed: 7)
  Failed tests:  1-7
  Non-zero exit status: 7
t/02-synopsis-proc.t (Wstat: 65280 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
t/03-synopsis-oo.t  (Wstat: 65280 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
t/04-proc-vs-oo.t   (Wstat: 65280 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
t/05-primitive.t    (Wstat: 65280 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
t/Crypt-Curve25519.t (Wstat: 512 Tests: 0 Failed: 0)
  Non-zero exit status: 2
  Parse errors: No plan found in TAP output
Files=7, Tests=12,  1 wallclock secs ( 0.06 usr  0.02 sys +  0.45 cusr  0.09 csys =  0.62 CPU)
Result: FAIL
Failed 7/7 test programs. 12/12 subtests failed.
make: *** [Makefile:1026: test_dynamic] Error 2
```

Ordinary builds  work ok:
```
perl Makefile.PL OPTIMIZE="-O3"
MakeMaker (v7.24)
Warning (non-fatal): Target 'dynamic' depends on targets in skipped section 'dynamic_lib'
Warning (non-fatal): Target 'static' depends on targets in skipped section 'static_lib'
Generating a Unix-style Makefile
Writing Makefile for Crypt::Curve25519::curve25519donna
Generating a Unix-style Makefile
Writing Makefile for Crypt::Curve25519
Writing MYMETA.yml and MYMETA.json
$ make
Skip blib/lib/Crypt/Curve25519.pm (unchanged)
cd curve25519-donna-master && make LIBPERL_A="libperl.a" LINKTYPE="dynamic" OPTIMIZE="-O3" PREFIX="/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot" PASTHRU_DEFINE=' ' PASTHRU_INC='-I. '
make[1]: Entering directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
cc -c -I.   -fstack-protector-strong -fno-stack-protector -DPERL_HASH_FUNC_SDBM -DPERL_DISABLE_PMC -fwrapv -fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O3   -DVERSION=\"\" -DXS_VERSION=\"\" -fPIC "-I/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/CORE"   curve25519-donna-c64.c
ar cr curve25519-donna-c64.a curve25519-donna-c64.o
: curve25519-donna-c64.a
make[1]: Leaving directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
make[1]: Entering directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
Running Mkbootstrap for Curve25519 ()
chmod 644 "Curve25519.bs"
"/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/bin/perl5.25.10" -MExtUtils::Command::MM -e 'cp_nonempty' -- Curve25519.bs blib/arch/auto/Crypt/Curve25519/Curve25519.bs 644
cc -c  -I. -fstack-protector-strong -fno-stack-protector -DPERL_HASH_FUNC_SDBM -DPERL_DISABLE_PMC -fwrapv -fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O3   -DVERSION=\"0.05\" -DXS_VERSION=\"0.05\" -fPIC "-I/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/lib/5.25.10/x86_64-linux/CORE"   Curve25519.c
rm -f blib/arch/auto/Crypt/Curve25519/Curve25519.so
cc  -shared  -fstack-protector-strong -fno-stack-protector -O3 -march=native -mtune=native -L/usr/local/lib Curve25519.o  -o blib/arch/auto/Crypt/Curve25519/Curve25519.so curve25519-donna-master/curve25519-donna-c64.a \
      \
  
chmod 755 blib/arch/auto/Crypt/Curve25519/Curve25519.so
$ make test
"/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/bin/perl5.25.10" -MExtUtils::Command::MM -e 'cp_nonempty' -- Curve25519.bs blib/arch/auto/Crypt/Curve25519/Curve25519.bs 644
make[1]: Entering directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
make[1]: Entering directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
No tests defined for Crypt::Curve25519::curve25519donna extension.
make[1]: Leaving directory '/home/kent/.cpanm/work/1490279048.28810/Crypt-Curve25519-0.05/curve25519-donna-master'
PERL_DL_NONLAZY=1 "/home/kent/perl5/perlbrew/perls/5.25.10-nossp-sdbm-nopmc-nodot/bin/perl5.25.10" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/00-compile.t ........ ok   
t/01-exceptions.t ..... ok   
t/02-synopsis-proc.t .. ok   
t/03-synopsis-oo.t .... ok   
t/04-proc-vs-oo.t ..... ok   
t/05-primitive.t ...... ok   
t/Crypt-Curve25519.t .. ok       
All tests successful.
Files=7, Tests=10028,  7 wallclock secs ( 1.29 usr  0.14 sys +  6.12 cusr  0.16 csys =  7.71 CPU)
Result: PASS
```

**Comment by ajgb** [comment]
Could you test just released version 0.06?

Thank you,
Alex

**Comment by ghost** [comment]
Crypt-Curve25519-0.06.tar.gz
In file included from Curve25519.xs:7:
curve25519-donna-c64.c:99:1: error: conflicting types for 'fmul'

fedora 29 - Just renamed to curve_fmul, installed manually and all is good.
